File: client/src/entities/dummy.ts
// A stationary target dummy for the M2 local combat sandbox. Render-only state
// lives here; the capsule it exposes (capsuleTarget) is what hitscan and the
// explosion sim test against. Damage is applied locally in M2; M3 moves the
// authority to the server. Box-humanoid (stacked boxes, enemy-red, PRD §10)
// with a billboarded HP bar that follows the camera.

import * as THREE from 'three';
import { TUNING, clamp } from '@rivals/shared';
import type { Vec3 } from '@rivals/shared';
import type { CapsuleTarget } from '../combat/hitscan';
import { Humanoid } from './humanoid';

// Capsule the analytic/explosion math sees. Slightly wider/shorter than the
// player capsule so the humanoid silhouette reads as a fair target.
const RADIUS = 0.4;
const HALF_HEIGHT = 0.5; // cylinder half-segment; full hittable height ~1.8

const MAX_HP = TUNING.combat.spawnHealth;
const RESET_DELAY = 2.0; // seconds after the last hit -> heal back to full
const DEATH_HIDE = 0.5; // seconds the body stays hidden after a kill
const FLASH_TIME = 0.12; // seconds the body glows white after a hit

const HP_BAR_WIDTH = 1.0;
const HP_BAR_HEIGHT = 0.12;
const HP_BAR_LIFT = 1.35; // meters above capsule center

const BODY_COLOR = 0xff5a5a; // enemy-red (PRD §10)
const FLASH_COLOR = 0xffffff;
const HP_GREEN = new THREE.Color(0x46d35a);
const HP_RED = new THREE.Color(0xff4040);

// Scratch reused every frame; the HP bar billboard only needs the camera's
// world position, so no per-frame allocation here.
const _camPos = new THREE.Vector3();

export class Dummy {
  readonly object: THREE.Object3D;
  readonly id: number;

  center: Vec3;
  radius = RADIUS;
  halfHeight = HALF_HEIGHT;
  facingYaw: number;
  hp = MAX_HP;

  private readonly body: Humanoid;
  private readonly bodyMat: THREE.MeshStandardMaterial;
  private readonly hpBar: THREE.Sprite;
  private readonly hpBarMat: THREE.SpriteMaterial;

  private flash = 0; // remaining flash time
  private resetTimer = 0; // counts down to heal-to-full (0 = idle)
  private deathTimer = 0; // counts down while dead/hidden (0 = alive)

  // Reuse one Color instance for HP-bar tinting to avoid per-frame allocation.
  private readonly _tint = new THREE.Color();

  constructor(id: number, center: Vec3) {
    this.id = id;
    this.center = { x: center.x, y: center.y, z: center.z };
    this.facingYaw = Math.PI; // face back toward the arena / spawn

    this.object = new THREE.Group();
    this.object.name = `dummy:${id}`;
    this.object.position.set(center.x, center.y, center.z);
    this.object.rotation.y = this.facingYaw;

    // Shared material for all body boxes so a single color write flashes them.
    this.bodyMat = new THREE.MeshStandardMaterial({
      color: BODY_COLOR,
      roughness: 0.85,
      metalness: 0.0,
      flatShading: true,
    });

    this.body = new Humanoid(this.bodyMat);
    this.object.add(this.body.object);

    // Billboarded HP bar. A SpriteMaterial with a canvas texture: the green/red
    // fill is drawn into the canvas and the texture updated only on change.
    this.hpBarMat = new THREE.SpriteMaterial({
      color: 0xffffff,
      depthTest: false,
      transparent: true,
    });
    this.hpBar = new THREE.Sprite(this.hpBarMat);
    this.hpBar.scale.set(HP_BAR_WIDTH, HP_BAR_HEIGHT, 1);
    this.hpBar.position.set(0, HP_BAR_LIFT, 0);
    this.hpBar.renderOrder = 999;
    this.object.add(this.hpBar);

    this.refreshHpBar();
  }

  capsuleTarget(): CapsuleTarget {
    return {
      id: this.id,
      center: this.center,
      radius: this.radius,
      halfHeight: this.halfHeight,
    };
  }

  applyDamage(amount: number): void {
    if (this.deathTimer > 0) return; // already dead; ignore until respawn
    this.hp = clamp(this.hp - amount, 0, MAX_HP);
    this.flash = FLASH_TIME;
    this.refreshHpBar();

    if (this.hp <= 0) {
      // Hide the body briefly, then respawn at full.
      this.deathTimer = DEATH_HIDE;
      this.resetTimer = 0;
      this.body.object.visible = false;
      this.hpBar.visible = false;
    } else {
      // (Re)start the heal-back timer on every non-fatal hit.
      this.resetTimer = RESET_DELAY;
    }
  }

  update(dt: number, camera: THREE.Camera): void {
    // --- death / respawn ---
    if (this.deathTimer > 0) {
      this.deathTimer -= dt;
      if (this.deathTimer <= 0) {
        this.deathTimer = 0;
        this.hp = MAX_HP;
        this.body.object.visible = true;
        this.hpBar.visible = true;
        this.flash = 0;
        this.bodyMat.color.setHex(BODY_COLOR);
        this.refreshHpBar();
      }
      return; // hidden: skip flash/billboard work
    }

    // --- heal-to-full timer ---
    if (this.resetTimer > 0) {
      this.resetTimer -= dt;
      if (this.resetTimer <= 0) {
        this.resetTimer = 0;
        this.hp = MAX_HP;
        this.refreshHpBar();
      }
    }

    // --- flash decay (lerp body color back to enemy-red) ---
    if (this.flash > 0) {
      this.flash -= dt;
      const t = clamp(this.flash / FLASH_TIME, 0, 1);
      this.bodyMat.color.setHex(BODY_COLOR).lerp(FLASH_WHITE, t);
      if (this.flash <= 0) {
        this.flash = 0;
        this.bodyMat.color.setHex(BODY_COLOR);
      }
    }

    // --- idle animation (stationary target → breathing only) ---
    this.body.update(dt, 0, true);

    // --- billboard the HP bar toward the camera (yaw only via Sprite) ---
    // THREE.Sprite already faces the camera; nothing to rotate. Keep the camera
    // reference used so the signature stays honest and future bars (text) can
    // orient. Touch camera world pos to avoid an unused-param lint and to allow
    // distance-based scaling later.
    camera.getWorldPosition(_camPos);
  }

  private refreshHpBar(): void {
    const frac = clamp(this.hp / MAX_HP, 0, 1);
    this._tint.copy(HP_RED).lerp(HP_GREEN, frac);
    this.hpBarMat.color.copy(this._tint);
    // Shrink the bar width to the fill fraction so it drains left-to-right.
    this.hpBar.scale.x = HP_BAR_WIDTH * Math.max(frac, 0.001);
    // Keep the (shrinking) bar left-anchored under the head.
    this.hpBar.position.x = -((HP_BAR_WIDTH - this.hpBar.scale.x) / 2);
  }
}

const FLASH_WHITE = new THREE.Color(FLASH_COLOR);

File: godot/src/Client/Render/Dummy.cs
// A stationary target dummy for the offline combat sandbox. Render-only state
// lives here; the capsule it exposes is what hitscan and the explosion sim test
// against. Ported from client/src/entities/dummy.ts.

using System;
using Godot;
using WebRivals.Shared;
using WebRivals.Client.Combat;
using static WebRivals.Client.GodotConv;

namespace WebRivals.Client.Render;

public sealed class Dummy
{
    public readonly Node3D Object;
    public readonly int Id;

    public Vec3 Center;
    public double Radius = 0.4;
    public double HalfHeight = 0.5;
    public double FacingYaw;
    public double Hp;

    private const double MaxHp = 100; // Tuning.Combat.SpawnHealth
    private const double ResetDelay = 2.0;
    private const double DeathHide = 0.5;
    private const double FlashTime = 0.12;

    private readonly Humanoid _body;
    private readonly StandardMaterial3D _bodyMat;
    private readonly MeshInstance3D _hpBar;
    private readonly StandardMaterial3D _hpMat;

    private double _flash = 0;
    private double _resetTimer = 0;
    private double _deathTimer = 0;

    private static readonly Color BodyColor = new(1f, 0.353f, 0.353f);
    private static readonly Color FlashWhite = new(1f, 1f, 1f);
    private static readonly Color HpGreen = new(0.275f, 0.827f, 0.353f);
    private static readonly Color HpRed = new(1f, 0.251f, 0.251f);

    public Dummy(int id, Vec3 center)
    {
        Id = id;
        Center = new Vec3(center.X, center.Y, center.Z);
        FacingYaw = Math.PI;
        Hp = MaxHp;

        Object = new Node3D { Name = $"Dummy{id}" };
        Object.Position = ToGd(center);
        Object.Rotation = new Vector3(0, (float)FacingYaw, 0);

        _bodyMat = new StandardMaterial3D { AlbedoColor = BodyColor, Roughness = 0.85f, Metallic = 0.0f };
        _body = new Humanoid(_bodyMat);
        Object.AddChild(_body.Object);

        _hpMat = new StandardMaterial3D
        {
            AlbedoColor = HpGreen,
            ShadingMode = BaseMaterial3D.ShadingModeEnum.Unshaded,
            BillboardMode = BaseMaterial3D.BillboardModeEnum.Enabled,
            NoDepthTest = true,
            RenderPriority = 10,
        };
        _hpBar = new MeshInstance3D
        {
            Mesh = new QuadMesh { Size = new Vector2(1.0f, 0.12f) },
            MaterialOverride = _hpMat,
            Position = new Vector3(0, 1.35f, 0),
            CastShadow = GeometryInstance3D.ShadowCastingSetting.Off,
        };
        Object.AddChild(_hpBar);
        RefreshHpBar();
    }

    public CapsuleTarget CapsuleTarget() => new() { Id = Id, Center = Center, Radius = Radius, HalfHeight = HalfHeight };

    public void ApplyDamage(double amount)
    {
        if (_deathTimer > 0) return;
        Hp = VMath.Clamp(Hp - amount, 0, MaxHp);
        _flash = FlashTime;
        RefreshHpBar();

        if (Hp <= 0)
        {
            _deathTimer = DeathHide;
            _resetTimer = 0;
            _body.Object.Visible = false;
            _hpBar.Visible = false;
        }
        else
        {
            _resetTimer = ResetDelay;
        }
    }

    public void Update(double dt)
    {
        if (_deathTimer > 0)
        {
            _deathTimer -= dt;
            if (_deathTimer <= 0)
            {
                _deathTimer = 0;
                Hp = MaxHp;
                _body.Object.Visible = true;
                _hpBar.Visible = true;
                _flash = 0;
                _bodyMat.AlbedoColor = BodyColor;
                RefreshHpBar();
            }
            return;
        }

        if (_resetTimer > 0)
        {
            _resetTimer -= dt;
            if (_resetTimer <= 0)
            {
                _resetTimer = 0;
                Hp = MaxHp;
                RefreshHpBar();
            }
        }

        if (_flash > 0)
        {
            _flash -= dt;
            double t = VMath.Clamp(_flash / FlashTime, 0, 1);
            _bodyMat.AlbedoColor = BodyColor.Lerp(FlashWhite, (float)t);
            if (_flash <= 0)
            {
                _flash = 0;
                _bodyMat.AlbedoColor = BodyColor;
            }
        }

        _body.Update(dt, 0, true);
    }

    public bool Visible
    {
        get => Object.Visible;
        set => Object.Visible = value;
    }

    private void RefreshHpBar()
    {
        double frac = VMath.Clamp(Hp / MaxHp, 0, 1);
        _hpMat.AlbedoColor = HpRed.Lerp(HpGreen, (float)frac);
        _hpBar.Scale = new Vector3((float)Math.Max(frac, 0.001), 1, 1);
    }
}
